### PR TITLE
LNS: Revise name validation on client side

### PR DESCRIPTION
**Pull request by Doy-lee**
In summary

```
SESSION or WALLET
Name has to start with a (alphanumeric or underscore), and can have
(alphanumeric, hyphens or underscores) in between and must end with
a (alphanumeric or underscore)

LOKINET
Domain has to start with an alphanumeric, and can have (alphanumeric or
hyphens) in between must end with a '.loki'
```

Also add a check to the LNS query command, to early disconnect the request if they request an invalid name via RPC.

### CODE REVIEW

File: src/cryptonote_core/loki_name_system.cpp [hunk 1]
```diff
@@ -439,57 +439,75 @@ bool validate_lns_name(mapping_type type, std::string const &name, std::string *
...
+  char const *char_preceeding_suffix = nullptr;
+  if (is_lokinet)
   {
-  }
-  else if (is_lokinet_type(type))
-  {
-    // Domain has to start with a letter or digit, and can have letters, digits, or hyphens in between and must end with a .loki
+
+    // LOKINET
+    // Domain has to start with an alphanumeric, and can have (alphanumeric or hyphens) in between must end with a '.loki'
```
Comment (jagerman) [inline src/cryptonote_core/loki_name_system.cpp]: Also the character preceding the ".loki" has to be alphanumeric.  (The code and regex comment do this, but it's missing from the comment description here).

File: src/cryptonote_core/loki_name_system.cpp [hunk 2]
```diff
@@ -439,57 +439,75 @@ bool validate_lns_name(mapping_type type, std::string const &name, std::string *
...
+      char_preceeding_suffix = name_suffix - 1;
+      if (check_condition(!char_is_alphanum(char_preceeding_suffix[0]), reason, "LNS type=", type ,", specifies mapping from name->value where the character preceeding the <char>.loki is not alphanumeric, char=", char_preceeding_suffix[0], ", name=", name))
         return false;
     }
+  }
+  else
+  {
+    // SESSION or WALLET
+    // Name has to start with a (alphanumeric or underscore), and can have (alphanumeric, hyphens or underscores) in between and must end with a (alphanumeric or underscore)
+    // ^[a-z0-9_]([a-z0-9-_]*[a-z0-9_])?
```
Comment (jagerman) [inline src/cryptonote_core/loki_name_system.cpp]: Technically should be anchored with a $ at the end

File: src/cryptonote_core/loki_name_system.cpp [hunk 3]
```diff
@@ -439,57 +439,75 @@ bool validate_lns_name(mapping_type type, std::string const &name, std::string *
...
+      // Must end with .loki
+      char const SUFFIX[]     = ".loki";
+      char const *name_suffix = name.data() + (name.size() - loki::char_count(SUFFIX));
+      if (check_condition((memcmp(name_suffix, SUFFIX, loki::char_count(SUFFIX)) != 0), reason, "LNS type=", type, ", specifies mapping from name->value where the name does not end with the domain .loki, name=", name))
+        return false;
+
+      // Characted preceeding suffix must be alphanumeric
+      char_preceeding_suffix = name_suffix - 1;
+      if (check_condition(!char_is_alphanum(char_preceeding_suffix[0]), reason, "LNS type=", type ,", specifies mapping from name->value where the character preceeding the <char>.loki is not alphanumeric, char=", char_preceeding_suffix[0], ", name=", name))
         return false;
```
Comment (jagerman) [inline src/cryptonote_core/loki_name_system.cpp]: Is this missing a check for [a-z0-9-] on the intermediate characters, or it is being handled somewhere else and I missed it?

Comment (jagerman) [reply src/cryptonote_core/loki_name_system.cpp]: Oh, it's below.

File: src/cryptonote_core/loki_name_system.cpp [hunk 4]
```diff
@@ -439,57 +439,75 @@ bool validate_lns_name(mapping_type type, std::string const &name, std::string *
...
 
-    // Characted preceeding suffix must be alphanumeric
-    char const *char_preceeding_suffix = name_suffix - 1;
-    if (check_condition(!char_is_alphanum(char_preceeding_suffix[0]), reason, "LNS type=lokinet, specifies mapping from name -> value where the character preceeding the <char>.loki is not alphanumeric, char=", char_preceeding_suffix[0], ", name=", name))
+  char const *begin = name.data() + 1;
+  char const *end   = char_preceeding_suffix;
+  for (char const *it = begin; it < end; it++) // Inbetween start and preceeding suffix, (alphanumeric, hyphen or underscore) characters permitted
+  {
+    char c = it[0];
+    if (check_condition(!(char_is_alphanum(c) || c == '-' || c == '_'), reason, "LNS type=", type, ", specifies mapping from name->value where the domain name contains more than the permitted alphanumeric or hyphen characters name=", name))
```
Comment (jagerman) [inline src/cryptonote_core/loki_name_system.cpp]: I think the `c == '_'` here should be `(type != lokinet && c == '_')` to disallow `_`'s from lokinet domains.

File: src/cryptonote_core/loki_name_system.cpp [hunk 5]
```diff
@@ -439,57 +439,75 @@ bool validate_lns_name(mapping_type type, std::string const &name, std::string *
...
+  char const *char_preceeding_suffix = nullptr;
+  if (is_lokinet)
   {
-  }
-  else if (is_lokinet_type(type))
-  {
-    // Domain has to start with a letter or digit, and can have letters, digits, or hyphens in between and must end with a .loki
+
+    // LOKINET
+    // Domain has to start with an alphanumeric, and can have (alphanumeric or hyphens) in between, the character before the suffix <char>'.loki' must be alphanumeric followed by the suffix '.loki'
```
Comment (jagerman) [inline src/cryptonote_core/loki_name_system.cpp]: We should also disallow "localhost.loki" as that's a builtin alias.  (The only other one lokinet has is "random.snode", but that doesn't matter here).

File: tests/unit_tests/loki_name_system.cpp [hunk 6]
```diff
@@ -3,63 +3,66 @@
...
+      {"a.loki", true},
+      {"domain.loki", true},
+      {"xn--tda.loki", true},
+      {"xn--Mchen-Ost-9db-u6b.loki", true},
+
+      {"abc.domain.loki", false},
+      {"a", false},
+      {"a.loko", false},
+      {"a domain name.loki", false},
+      {"-.loki", false},
```
Comment (jagerman) [inline tests/unit_tests/loki_name_system.cpp]: `{"a_b.loki", false},`

File: src/cryptonote_core/loki_name_system.cpp (hunk 5 above)
Comment (Doy-lee) [reply src/cryptonote_core/loki_name_system.cpp]: Added

File: src/cryptonote_core/loki_name_system.cpp (hunk 4 above)
Comment (Doy-lee) [reply src/cryptonote_core/loki_name_system.cpp]: I've separated the cases for clarity

File: tests/unit_tests/loki_name_system.cpp (hunk 6 above)
Comment (Doy-lee) [reply tests/unit_tests/loki_name_system.cpp]: Added

**Comment by Doy-lee** [comment]
Added lowercasing of lokinet names

**Comment by jagerman** [comment]
Looks good to me.  The only issue I'm unsure of is whether case insensitivity should apply to wallet/session names.  (It *has* to apply to lokinet since lokinet lookups go through DNS, but non-lokinet mappings don't have to be).

**Comment by Doy-lee** [comment]
Re-updated to lowercase all names

File: src/cryptonote_core/loki_name_system.cpp [hunk 7]
```diff
@@ -463,63 +463,87 @@ static bool check_condition(bool condition, std::string* reason, T&&... args) {
...
-  else if (is_lokinet_type(type))         max_name_len = lns::LOKINET_DOMAIN_NAME_MAX;
   else
   {
     if (reason) err_stream << "LNS type=" << type << ", specifies unhandled mapping type in name validation";
     return false;
   }
 
   // NOTE: Validate name length
-  if (check_condition((name.empty() || name.size() > max_name_len), reason, "LNS type=", type, ", specifies mapping from name -> value where the name's length=", name.size(), " is 0 or exceeds the maximum length=", max_name_len, ", given name=", name))
+  name = tools::lowercase_ascii_string(name);
```
Comment (jagerman) [inline src/cryptonote_core/loki_name_system.cpp]: If you use `name = lowercase(std::move(name));` here you can save a string reallocation.

Comment (Doy-lee) [reply src/cryptonote_core/loki_name_system.cpp]: Added

**Pull request merged**